server/protocols/jmap/src/main/java/org/apache/james/jmap/model/UpdateMessagePatch.java

```diff
@@ -81,14 +90,27 @@ public UpdateMessagePatch build() {
                     .message("mailboxIds property is not supposed to be empty")
                     .build()));
             }
-            return new UpdateMessagePatch(mailboxIds, isUnread, isFlagged, isAnswered, ImmutableList.copyOf(validationResult));
+            if (!noConflict()) {
+                validationResult(ImmutableSet.of(ValidationResult.builder()
+                    .property("keywords")
+                    .message("Does not support updating both is* and keywords")
+                    .build()));
+            }
+            return new UpdateMessagePatch(mailboxIds, isUnread, isFlagged, isAnswered, keywords, ImmutableList.copyOf(validationResult));
         }
+
+        private boolean noConflict() {
+            return !keywords.isPresent()
+                    || (!isAnswered.isPresent() && !isFlagged.isPresent() && !isUnread.isPresent());
```

Comment:
I see this pattern of code too much in this review, you lack an abstraction IMO.